File: utils/states.js
const ora = require("ora");
const axios = require("axios");
const fs = require('fs');
const os = require('os');
const FormData = require('form-data');
const handelError = require("alerts-in-cli");
const BASE_SERVER_URL = "https://occt.in"

let spinner = ora({ text: '' });

const uploadingFileCompile = async (flags) => {
    spinner.start(`Uploading file and processing...`);
    try {
        const formData = new FormData();
        const os_info = os.platform() + " " + os.arch();
        if(os_info === "darwin arm64") {
            formData.append('os', "macos-arm64");
        }else if(os_info === "darwin x64") {
            formData.append('os', "macos-x64");
        }else if(os_info === "win32 x64") {
            formData.append('os', "windows-x64");
        }else if(os_info === "linux x64") {
            formData.append('os', "linux-x64");
        }else {
            formData.append('os', os_info);
        }
        formData.append('file', fs.createReadStream(flags.compile), flags.compile);
        if(flags.outputFile) {
            formData.append('outputName', flags.outputFile);
        }

        if (flags.args) {
            formData.append('args', JSON.stringify(flags.args));
        }

        const response = await axios.post(`${BASE_SERVER_URL}/compile`, formData, {
            headers: formData.getHeaders()
        });

        if (response.status !== 200) {
            console.error('Compilation failed:', response.statusText);
            process.exit(1);
        }

        if(response.status === 200) {
            // console.log('✅ Compilation result:', response.data);

            const downloadUrl = `${BASE_SERVER_URL}${response.data.download}`;
            const writer = fs.createWriteStream(response.data.download.replace("/download/", ""));

            const download = await axios.get(downloadUrl, { responseType: 'stream' });
            download.data.pipe(writer);

            writer.on('finish', () => console.log(`Compiled file: ${response.data.download.replace("/download/", "")}`));
        }
        spinner.stop();
    } catch (error) {
        spinner.fail();
        handelError({
            type: "error",
            msg: error.response.data.error || error,
            name: "API Call fail"
        });
        process.exit(1);
    }
}

const uploadingFileInterprate = async (flags, cli = {}) => {
    spinner.start(`Uploading file and processing...`);
    try {
        const formData = new FormData();
        formData.append('file', fs.createReadStream(flags.interpreted), flags.interpreted);
        if(flags.outputFile) {
            formData.append('outputName', flags.outputFile);
        }

        if (flags.args) {
            if((cli.input.length > 0) && (cli.flags.args.length > 0)) {
                flags.args += ` ${cli.input.join(' ')}`;
            }
            formData.append('args', JSON.stringify(flags.args + ''));
        }

        const response = await axios.post(`${BASE_SERVER_URL}/interprate`, formData, {
            headers: formData.getHeaders()
        });

        if (response.status !== 200) {
            console.error('Compilation failed:', response.statusText);
            process.exit(1);
        }

        if(response.status === 200) {
            // console.log('✅ Compilation result:', response.data);

            console.log(`\n---\n${response.data.output}---`);
        }
        spinner.stop();
    } catch (error) {
        spinner.fail();
        handelError({
            type: "error",
            msg: error.response.data.error || error,
            name: "API Call fail"
        });
        process.exit(1);
    }
}

module.exports = {
    uploadingFileCompile,
    uploadingFileInterprate
}

File: occt/states.py
import requests
import sys
import json
from pathlib import Path
from yaspin import yaspin

BASE_SERVER_URL = "http://localhost:8000"

def handle_error(error, name="API Call fail"):
    print(f"\n[ERROR] {name}: {error}")
    sys.exit(1)


def uploading_file_compile(flags):
    with yaspin(text="Uploading file and processing...", color="yellow") as spinner:
        try:
            file_path = Path(flags.get("compile"))
            if not file_path.exists():
                handle_error("File not found")

            files = {"file": (file_path.name, open(file_path, "rb"))}
            data = {}

            if flags.get("outputFile"):
                data["outputName"] = flags["outputFile"]

            if flags.get("args"):
                data["args"] = json.dumps(flags["args"])

            response = requests.post(f"{BASE_SERVER_URL}/compile", files=files, data=data)

            if response.status_code != 200:
                handle_error(response.text)

            resp_data = response.json()
            download_url = f"{BASE_SERVER_URL}{resp_data['download']}"
            output_file = resp_data["download"].replace("/download/", "")

            download = requests.get(download_url, stream=True)
            with open(output_file, "wb") as f:
                for chunk in download.iter_content(1024):
                    f.write(chunk)

            spinner.ok("✅")
            print(f"Compiled file: {output_file}")

        except Exception as e:
            spinner.fail("💥")
            handle_error(e)


def uploading_file_interprate(flags, cli=None):
    cli = cli or {}
    with yaspin(text="Uploading file and processing...", color="yellow") as spinner:
        try:
            file_path = Path(flags.get("interpreted"))
            if not file_path.exists():
                handle_error("File not found")

            files = {"file": (file_path.name, open(file_path, "rb"))}
            data = {}

            if flags.get("outputFile"):
                data["outputName"] = flags["outputFile"]

            if flags.get("args"):
                args_val = flags["args"]
                if cli.get("input") and cli.get("flags", {}).get("args"):
                    args_val += " " + " ".join(cli["input"])
                data["args"] = json.dumps(str(args_val))

            response = requests.post(f"{BASE_SERVER_URL}/interprate", files=files, data=data)

            if response.status_code != 200:
                handle_error(response.text)

            resp_data = response.json()
            spinner.ok("✅")
            print(f"\n---\n{resp_data['output']}---")

        except Exception as e:
            spinner.fail("💥")
            handle_error(e)
